mgr/dashboard: Cloning subvolume not listing _nogroup if no subvolume

dnyanee1997 opened the pull request:
Fixes: https://tracker.ceph.com/issues/67891

Signed-off-by: Dnyaneshwari talwekar <dtalweka@redhat.com>

Tracker: https://tracker.ceph.com/issues/67891

![Screenshot from 2024-09-10 17-16-15](https://github.com/user-attachments/assets/9bebd709-abcd-43df-a89d-61e2ea9d5505)


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-snapshots-list/cephfs-subvolume-snapshots-list.component.ts:
```diff
@@ -251,6 +251,17 @@ export class CephfsSubvolumeSnapshotsListComponent implements OnInit, OnChanges
     const allGroups = Array.from(this.subvolumeGroupList).map((group) => {
       return { value: group, text: group === '' ? '_nogroup' : group };
     });
+    const nogroupExists = allGroups.some((group) => group.text === '_nogroup');
```
Lets create a constant for this in `src/app/shared/constants` , we can have a file created `cephfs.ts`.
DEFAULT_SUBVOLUME_GROUP='_nogroup' and export from there all places, in your NFS PR as well.

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-snapshots-list/cephfs-subvolume-snapshots-list.component.ts:
```diff
@@ -251,6 +251,17 @@ export class CephfsSubvolumeSnapshotsListComponent implements OnInit, OnChanges
     });
+    const nogroupExists = allGroups.some((group) => group.text === '_nogroup');
+
+    if (!nogroupExists) {
+      allGroups.unshift({ value: '', text: '_nogroup' });
```
Why are we adding "_nogroup" here ?

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/constants/cephfs.ts:
```diff
@@ -0,0 +1 @@
+export const DEFAULT_SUBVOLUME_GROUP = '_nogroup';
```
nit: missing newline

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-group/cephfs-subvolume-group.component.ts:
```diff
@@ -142,7 +142,21 @@ export class CephfsSubvolumeGroupComponent implements OnInit, OnChanges {
+            {
+              name: '_nogroup',
+              info: {
+                mode: 1,
+                bytes_pcent: 1,
+                bytes_quota: 1,
```
Why do you want to have mode as 1 and utilization as 1 for this? Isn't this wrong data?

We can try to ignore these and put N/A or `-` instead  
```suggestion
                mode: 1,
                bytes_pcent: 1,
                bytes_quota: 1,
```

cloudbehl replied:
@nizamial09 thoughts?

afreen23 replied:
Lets avoid adding these as discusses, we can have a separate PR/tracker for this if req. Thanks__

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-snapshots-list/cephfs-subvolume-snapshots-list.component.ts:
```diff
@@ -249,8 +250,19 @@ export class CephfsSubvolumeSnapshotsListComponent implements OnInit, OnChanges
-      return { value: group, text: group === '' ? '_nogroup' : group };
+      return { value: group, text: group === '' ? DEFAULT_SUBVOLUME_GROUP : group };
     });
+    const nogroupExists = allGroups.some((group) => group.text === DEFAULT_SUBVOLUME_GROUP);
+
+    if (!nogroupExists) {
+      allGroups.unshift({ value: '', text: DEFAULT_SUBVOLUME_GROUP });
+    } else {
+      allGroups.sort((a, b) => {
+        if (a.text === DEFAULT_SUBVOLUME_GROUP) return -1;
+        if (b.text === DEFAULT_SUBVOLUME_GROUP) return 1;
+        return 0;
+      });
+    }
```
We can have an option in html for this, choose and select that.
Its very similar to how NFS form lists "no_group" for Subvolume group field.
Lets copy the same approach here.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-snapshots-list/cephfs-subvolume-snapshots-list.component.ts:
```diff
@@ -249,8 +250,19 @@ export class CephfsSubvolumeSnapshotsListComponent implements OnInit, OnChanges
   cloneModal() {
     const cloneName = `clone_${moment().toISOString(true)}`;
     const allGroups = Array.from(this.subvolumeGroupList).map((group) => {
```
@dnyanee1997 one of the larger issue here is that `this.subvolumeGroupList` only has the list of subvolume groups which has subvolumes present inside it. You can see the filtration from the above codes.
```
this.cephfsSubvolumeService.existsInFs(this.fsName, group.name).pipe(
              switchMap((resp) => {
                if (resp) {
                  this.subvolumeGroupList.push(group.name);
                }
                return of(resp); // Emit the response
              })
            )
```

and this means that even if we fix the `_nogroup` scenario, all the other groups that doesn't have a subvolume is not listed here which shouldn't be the case. this dropdown should have the list of all the groups including the `_nogroup`. Could you look into that too?

dnyanee1997 commented:
@afreen23  As this is an modal, we don't have a option in html select and adding it from ts file. I have removed the sorting logic and added one line code. I think this will resolve our issue. Thanks.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-snapshots-list/cephfs-subvolume-snapshots-list.component.ts:
```diff
@@ -123,26 +124,11 @@ export class CephfsSubvolumeSnapshotsListComponent implements OnInit, OnChanges
-          groups.unshift({ name: '' });
-
-          const observables = groups.map((group) =>
-            this.cephfsSubvolumeService.existsInFs(this.fsName, group.name).pipe(
-              switchMap((resp) => {
-                if (resp) {
-                  this.subvolumeGroupList.push(group.name);
-                }
-                return of(resp); // Emit the response
-              })
-            )
-          );
```
maybe i was confusing before. we shouldn't remove this logic because this is still valid in the case of listing the snapshots. we don't want to list the groups that doesn't have a subvolume in the snapshot list page.

but in the form, we do need to list them. so the goal is to maintain two lists, one where it has all groups and the other where it has groups that has subvolumes. and preferrably we can populate these two inside a single subscription so we don't need to subscribe again. WDYT? @dnyanee1997

cloudbehl reviewed (approved):
LGTM, tested functionality at work

nmunet commented:
@dnyanee1997 , It seems like make check is failing, you need to run `npm run fix` to fix it.

nmunet commented:
![image](https://github.com/user-attachments/assets/d01a23f4-4944-43b0-bc4c-b1f84955177d)
Please fix these errors, else make check will fail again!

dnyanee1997 commented:
jenkins test api

dnyanee1997 commented:
jenkins test dashboard

dnyanee1997 commented:
jenkins test dashboard cephadm

cloudbehl commented:
Looks like a related error:

https://jenkins.ceph.com/job/ceph-dashboard-cephadm-e2e/12457/testReport/junit/(root)/should%20edit%20an%20export/nfsExport_page_Create__edit_and_delete_should_edit_an_export/

dnyanee1997 commented:
jenkins test dashboard cephadm

dnyanee1997 commented:
jenkins test dashboard

dnyanee1997 commented:
jenkins test dashboard cephadm

dnyanee1997 commented:
jenkins test dashboard

dnyanee1997 commented:
jenkins test dashboard

dnyanee1997 commented:
jenkins test dashboard cephadm

nizamial09 commented:
ceph dashboard e2e known failure on multi-cluster

[Test Result](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/16415/testReport/) (4 failures / -2)
[should switch to the second cluster and back to hub.Muti-cluster management page should switch to the second cluster and back to hub](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/16415/testReport/junit/(root)/should%20switch%20to%20the%20second%20cluster%20and%20back%20to%20hub/Muti_cluster_management_page_should_switch_to_the_second_cluster_and_back_to_hub/)
[should reconnect the second cluster.Muti-cluster management page should reconnect the second cluster](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/16415/testReport/junit/(root)/should%20reconnect%20the%20second%20cluster/Muti_cluster_management_page_should_reconnect_the_second_cluster/)
[should edit the second cluster.Muti-cluster management page should edit the second cluster](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/16415/testReport/junit/(root)/should%20edit%20the%20second%20cluster/Muti_cluster_management_page_should_edit_the_second_cluster/)
[should disconnect the second cluster.Muti-cluster management page should disconnect the second cluster](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/16415/testReport/junit/(root)/should%20disconnect%20the%20second%20cluster/Muti_cluster_management_page_should_disconnect_the_second_cluster/)

nizamial09 commented:
dashboard cephadm e2e was stuck which is being fixed through: https://github.com/ceph/ceph/pull/59691. so considering nothing blocking it, going to merge this one
